### ISSUE TITLE
Task 7: Implement model interpretability and explanation methods

### ISSUE THREAD

**Issue by jpequegn** [jpequegn/ts-rs#53]
# Task 7: Implement Model Interpretability and Explanation Methods

**Related to Epic**: #46  
**Depends on**: #47, #48, #49, #50, #52

## Objective
Implement comprehensive interpretability and explanation methods for machine learning models to help users understand model decisions, identify important features, and build trust in ML-powered time series analysis.

## Scope
- Create `src/ml/interpretability.rs` with explanation methods
- Implement feature importance analysis for time series models
- Add attention visualization for Transformer models
- Support counterfactual explanations and what-if analysis

### Core Functions to Implement
```rust
// Main interpretability interface
pub fn explain_model_prediction(
    model: &dyn ForecastMethod,
    time_series: &TimeSeries,
    explanation_config: &ExplanationConfig,
) -> Result<ModelExplanation, MLError>;

// Feature importance analysis
pub fn compute_feature_importance(
    model: &dyn ForecastMethod,
    data: &TimeSeries,
    method: ImportanceMethod,
) -> Result<FeatureImportance, MLError>;

// Attention analysis for Transformer models
pub fn analyze_attention_patterns(
    transformer: &TransformerForecaster,
    input_sequence: &TimeSeries,
) -> Result<AttentionAnalysis, MLError>;

// Counterfactual explanations
pub fn generate_counterfactual_explanations(
    model: &dyn ForecastMethod,
    original_series: &TimeSeries,
    explanation_config: &CounterfactualConfig,
) -> Result<Vec<CounterfactualExplanation>, MLError>;
```

### Data Structures
```rust
pub struct ModelExplanation {
    pub prediction: f64,
    pub feature_importance: FeatureImportance,
    pub temporal_importance: TemporalImportance,
    pub attention_weights: Option<AttentionWeights>,
    pub counterfactuals: Option<Vec<CounterfactualExplanation>>,
    pub confidence_explanation: ConfidenceExplanation,
    pub explanation_metadata: ExplanationMetadata,
}

pub struct ExplanationConfig {
    pub methods: Vec<ExplanationMethod>,
    pub baseline_strategy: BaselineStrategy,
    pub perturbation_config: PerturbationConfig,
    pub visualization_config: VisualizationConfig,
    pub confidence_level: f64,
}

pub enum ExplanationMethod {
    SHAP { background_samples: usize },
    LIME { neighborhood_size: usize, feature_selection: usize },
    IntegratedGradients { steps: usize, baseline: BaselineStrategy },
    GradientXInput,
    Occlusion { window_size: usize, stride: usize },
    PermutationImportance { n_repeats: usize },
    AttentionWeights,
    LayerWiseRelevance { rule: LRPRule },
}

pub struct FeatureImportance {
    pub temporal_importance: Vec<f64>,  // Importance per timestep
    pub feature_importance: HashMap<String, f64>,  // Importance per feature
    pub interaction_effects: Option<InteractionMatrix>,
    pub uncertainty: Option<Vec<f64>>,
    pub method_used: ImportanceMethod,
}

pub struct AttentionAnalysis {
    pub attention_maps: Vec<AttentionMap>,  // Per layer/head
    pub aggregated_attention: Vec<f64>,     // Aggregated across heads
    pub temporal_focus: TemporalFocus,
    pub attention_entropy: Vec<f64>,        // Attention distribution entropy
    pub head_specialization: HeadSpecialization,
}
```

### 1. Feature Attribution Methods
```rust
// SHAP (SHapley Additive exPlanations) implementation
pub struct SHAPExplainer {
    pub background_samples: usize,
    pub sampling_strategy: SamplingStrategy,
    pub coalition_sampling: CoalitionSampling,
    pub approximation_method: ShapApproximation,
}

impl SHAPExplainer {
    pub fn explain_prediction(&self, model: &dyn ForecastMethod, instance: &TimeSeries) -> Result<SHAPValues, MLError> {
        match self.approximation_method {
            ShapApproximation::Exact => self.compute_exact_shap_values(model, instance),
            ShapApproximation::Sampling { n_samples } => self.compute_sampling_shap_values(model, instance, n_samples),
            ShapApproximation::KernelSHAP => self.compute_kernel_shap_values(model, instance),
        }
    }
    
    fn compute_marginal_contributions(&self, model: &dyn ForecastMethod, coalitions: &[Coalition], instance: &TimeSeries) -> Result<Vec<f64>, MLError> {
        // Compute marginal contributions for each feature coalition
    }
}

// LIME (Local Interpretable Model-agnostic Explanations)
pub struct LIMEExplainer {
    pub neighborhood_size: usize,
    pub perturbation_strategy: PerturbationStrategy,
    pub local_model: LocalModelType,
    pub feature_selection_method: FeatureSelectionMethod,
}

impl LIMEExplainer {
    pub fn explain_prediction(&self, model: &dyn ForecastMethod, instance: &TimeSeries) -> Result<LIMEExplanation, MLError> {
        let neighborhood = self.generate_neighborhood(instance)?;
        let predictions = self.predict_neighborhood(model, &neighborhood)?;
        let local_model = self.fit_local_model(&neighborhood, &predictions)?;
        self.extract_feature_importance(&local_model)
    }
    
    fn generate_neighborhood(&self, instance: &TimeSeries) -> Result<Vec<TimeSeries>, MLError> {
        // Generate perturbed instances around the original time series
    }
}
```

### 2. Gradient-Based Attribution
```rust
// Integrated Gradients implementation
pub struct IntegratedGradients {
    pub steps: usize,
    pub baseline_strategy: BaselineStrategy,
    pub noise_tunnel: Option<NoiseTunnel>,
}

pub enum BaselineStrategy {
    Zero,
    Mean,
    Gaussian { mean: f64, std: f64 },
    Historical { lookback_periods: usize },
    Custom(TimeSeries),
}

impl IntegratedGradients {
    pub fn compute_attributions(&self, model: &dyn NeuralForecastMethod, input: &TimeSeries) -> Result<Vec<f64>, MLError> {
        let baseline = self.generate_baseline(input)?;
        let mut integrated_gradients = vec![0.0; input.values.len()];
        
        for i in 0..=self.steps {
            let alpha = i as f64 / self.steps as f64;
            let interpolated = self.interpolate(&baseline, input, alpha)?;
            let gradients = model.compute_gradients(&interpolated)?;
            
            for (j, grad) in gradients.iter().enumerate() {
                integrated_gradients[j] += grad / self.steps as f64;
            }
        }
        
        // Scale by input difference
        for (i, attr) in integrated_gradients.iter_mut().enumerate() {
            *attr *= input.values[i] - baseline.values[i];
        }
        
        Ok(integrated_gradients)
    }
}

// Gradient × Input attribution
pub struct GradientXInput;

impl GradientXInput {
    pub fn compute_attributions(&self, model: &dyn NeuralForecastMethod, input: &TimeSeries) -> Result<Vec<f64>, MLError> {
        let gradients = model.compute_gradients(input)?;
        let attributions = gradients.iter()
            .zip(input.values.iter())
            .map(|(grad, val)| grad * val)
            .collect();
        Ok(attributions)
    }
}
```

### 3. Attention Visualization and Analysis
```rust
pub struct AttentionAnalyzer {
    pub aggregation_method: AttentionAggregation,
    pub head_selection: HeadSelection,
    pub temporal_analysis: bool,
}

pub enum AttentionAggregation {
    Mean,
    Max,
    WeightedSum { weights: Vec<f64> },
    L2Norm,
}

pub enum HeadSelection {
    All,
    Top { k: usize },
    ByEntropy { threshold: f64 },
    Manual { head_indices: Vec<usize> },
}

impl AttentionAnalyzer {
    pub fn analyze_transformer_attention(&self, transformer: &TransformerForecaster, input: &TimeSeries) -> Result<AttentionAnalysis, MLError> {
        let attention_weights = transformer.extract_attention_weights(input)?;
        
        let aggregated_attention = self.aggregate_attention_heads(&attention_weights)?;
        let temporal_focus = self.analyze_temporal_patterns(&aggregated_attention)?;
        let attention_entropy = self.compute_attention_entropy(&attention_weights)?;
        let head_specialization = self.analyze_head_specialization(&attention_weights)?;
        
        Ok(AttentionAnalysis {
            attention_maps: attention_weights,
            aggregated_attention,
            temporal_focus,
            attention_entropy,
            head_specialization,
        })
    }
    
    pub fn visualize_attention_patterns(&self, analysis: &AttentionAnalysis) -> Result<AttentionVisualization, MLError> {
        // Create visualization data for attention patterns
    }
}

pub struct HeadSpecialization {
    pub head_roles: Vec<AttentionRole>,
    pub specialization_scores: Vec<f64>,
    pub interaction_patterns: InteractionMatrix,
}

pub enum AttentionRole {
    LocalPatterns,      // Focuses on nearby timesteps
    GlobalTrends,       // Captures long-term dependencies
    SeasonalPatterns,   // Detects seasonal cycles
    Anomalies,          // Highlights unusual patterns
    FeatureInteraction, // Models feature interactions
    Unknown,            // No clear specialization
}
```

### 4. Counterfactual Explanations
```rust
pub struct CounterfactualGenerator {
    pub search_method: CounterfactualSearch,
    pub constraints: CounterfactualConstraints,
    pub distance_metric: DistanceMetric,
    pub validity_check: ValidityChecker,
}

pub enum CounterfactualSearch {
    GradientBased { learning_rate: f64, max_iterations: usize },
    GeneticAlgorithm { population_size: usize, generations: usize },
    SimulatedAnnealing { initial_temp: f64, cooling_rate: f64 },
    NeighborhoodSearch { radius: f64, max_candidates: usize },
}

pub struct CounterfactualConstraints {
    pub feature_bounds: HashMap<String, (f64, f64)>,
    pub temporal_constraints: TemporalConstraints,
    pub realism_constraints: RealismConstraints,
    pub minimal_change_constraint: bool,
}

impl CounterfactualGenerator {
    pub fn generate_counterfactuals(&self, model: &dyn ForecastMethod, original: &TimeSeries, target_change: f64) -> Result<Vec<CounterfactualExplanation>, MLError> {
        let mut candidates = Vec::new();
        
        match &self.search_method {
            CounterfactualSearch::GradientBased { learning_rate, max_iterations } => {
                candidates = self.gradient_based_search(model, original, target_change, *learning_rate, *max_iterations)?;
            }
            CounterfactualSearch::GeneticAlgorithm { population_size, generations } => {
                candidates = self.genetic_algorithm_search(model, original, target_change, *population_size, *generations)?;
            }
            // ... other search methods
        }
        
        // Filter and rank counterfactuals
        let valid_counterfactuals = self.filter_valid_counterfactuals(candidates)?;
        let ranked_counterfactuals = self.rank_by_quality(&valid_counterfactuals)?;
        
        Ok(ranked_counterfactuals)
    }
}

pub struct CounterfactualExplanation {
    pub original_prediction: f64,
    pub counterfactual_prediction: f64,
    pub counterfactual_series: TimeSeries,
    pub changes_made: Vec<FeatureChange>,
    pub distance_from_original: f64,
    pub plausibility_score: f64,
    pub explanation_text: String,
}
```

### 5. Model-Specific Interpretability
```rust
// LSTM/GRU specific interpretability
pub struct RecurrentModelExplainer {
    pub cell_state_analysis: bool,
    pub gate_activation_analysis: bool,
    pub temporal_contribution_analysis: bool,
}

impl RecurrentModelExplainer {
    pub fn explain_lstm_prediction(&self, lstm: &LSTMForecaster, input: &TimeSeries) -> Result<LSTMExplanation, MLError> {
        let cell_states = lstm.extract_cell_states(input)?;
        let gate_activations = lstm.extract_gate_activations(input)?;
        let temporal_contributions = self.compute_temporal_contributions(lstm, input)?;
        
        Ok(LSTMExplanation {
            cell_state_evolution: cell_states,
            gate_importance: gate_activations,
            temporal_contributions,
            forget_gate_analysis: self.analyze_forget_gates(&gate_activations),
            input_gate_analysis: self.analyze_input_gates(&gate_activations),
        })
    }
}

// Ensemble model interpretability
pub struct EnsembleExplainer {
    pub model_contribution_analysis: bool,
    pub disagreement_analysis: bool,
    pub weight_evolution_analysis: bool,
}

impl EnsembleExplainer {
    pub fn explain_ensemble_prediction(&self, ensemble: &EnsembleForecaster, input: &TimeSeries) -> Result<EnsembleExplanation, MLError> {
        let individual_predictions = ensemble.get_individual_predictions(input)?;
        let model_weights = ensemble.get_current_weights();
        let model_contributions = self.compute_model_contributions(&individual_predictions, &model_weights)?;
        let disagreement_analysis = self.analyze_model_disagreement(&individual_predictions)?;
        
        Ok(EnsembleExplanation {
            model_contributions,
            weight_distribution: model_weights,
            disagreement_metrics: disagreement_analysis,
            consensus_confidence: self.compute_consensus_confidence(&individual_predictions),
        })
    }
}
```

### 6. Global Model Interpretability
```rust
pub struct GlobalModelAnalyzer {
    pub feature_interaction_analysis: bool,
    pub model_behavior_analysis: bool,
    pub decision_boundary_analysis: bool,
}

impl GlobalModelAnalyzer {
    pub fn analyze_global_behavior(&self, model: &dyn ForecastMethod, dataset: &[TimeSeries]) -> Result<GlobalModelAnalysis, MLError> {
        let feature_effects = self.compute_global_feature_effects(model, dataset)?;
        let interaction_effects = self.compute_feature_interactions(model, dataset)?;
        let model_complexity = self.estimate_model_complexity(model, dataset)?;
        let decision_regions = self.analyze_decision_regions(model, dataset)?;
        
        Ok(GlobalModelAnalysis {
            feature_effects,
            interaction_effects,
            model_complexity,
            decision_regions,
            summary_statistics: self.compute_summary_statistics(dataset),
        })
    }
}
```

### 1. Forecasting Integration
```rust
// Extend ForecastResult to include explanations
pub struct ExplainableForecastResult {
    pub forecast_result: ForecastResult,
    pub explanation: ModelExplanation,
    pub uncertainty_explanation: UncertaintyExplanation,
}

impl ForecastMethod for ExplainableForecaster {
    fn forecast(&self, data: &TimeSeries, horizon: usize) -> Result<ForecastResult, MLError> {
        let prediction = self.base_model.forecast(data, horizon)?;
        let explanation = self.explain_prediction(data, &prediction)?;
        
        Ok(ForecastResult {
            predictions: prediction.predictions,
            confidence_intervals: prediction.confidence_intervals,
            explanation: Some(explanation),
        })
    }
}
```

### 2. Anomaly Detection Integration
```rust
// Explainable anomaly detection
pub struct ExplainableAnomalyDetector {
    pub base_detector: Box<dyn AnomalyDetector>,
    pub explanation_method: AnomalyExplanationMethod,
}

impl ExplainableAnomalyDetector {
    pub fn detect_and_explain(&self, data: &TimeSeries) -> Result<ExplainableAnomalyResult, MLError> {
        let anomalies = self.base_detector.detect_anomalies(data)?;
        let explanations = anomalies.iter()
            .map(|anomaly| self.explain_anomaly(data, anomaly))
            .collect::<Result<Vec<_>, _>>()?;
            
        Ok(ExplainableAnomalyResult {
            anomalies,
            explanations,
        })
    }
}
```

### 1. Explanation Visualization
```rust
pub struct ExplanationVisualizer {
    pub plot_config: PlotConfig,
    pub interactive_features: InteractiveConfig,
}

impl ExplanationVisualizer {
    pub fn create_feature_importance_plot(&self, importance: &FeatureImportance) -> Result<PlotData, MLError> {
        // Create visualization for feature importance
    }
    
    pub fn create_attention_heatmap(&self, attention: &AttentionAnalysis) -> Result<PlotData, MLError> {
        // Create heatmap visualization for attention weights
    }
    
    pub fn create_counterfactual_comparison(&self, original: &TimeSeries, counterfactual: &CounterfactualExplanation) -> Result<PlotData, MLError> {
        // Create side-by-side comparison of original and counterfactual
    }
}
```

## Acceptance Criteria
- [ ] Feature attribution methods produce meaningful importance scores
- [ ] Attention visualization clearly shows model focus patterns
- [ ] Counterfactual explanations provide realistic alternative scenarios
- [ ] Model-specific explanations capture unique characteristics of each architecture
- [ ] Global analysis provides insights into overall model behavior
- [ ] Integration with forecasting and anomaly detection works seamlessly
- [ ] Explanations are consistent across multiple runs (deterministic when possible)
- [ ] Comprehensive test suite validates explanation quality and correctness

## Test Cases Required
- Feature importance correlation with known ground truth
- Attention pattern analysis on synthetic sequences with known dependencies
- Counterfactual generation quality and realism
- Explanation consistency across model types
- Performance benchmarks for explanation generation time
- Integration testing with all ML models

## Dependencies
- Tasks 1-6 results for all ML models and infrastructure
- Visualization libraries for explanation plots
- Optimization libraries for counterfactual search
- Statistical libraries for importance testing

## Estimated Time
18-22 hours

## Next Steps
After completion, enables:
- Task 8: Real-time model monitoring and adaptation
- Complete ML pipeline with full interpretability
- Trustworthy AI implementation for time series analysis
- Advanced debugging and model improvement capabilities
